cranelift/interpreter/src/step.rs

```diff
@@ -668,11 +668,29 @@ where
         Opcode::IshlImm => binary(Value::shl, arg(0)?, imm_as_ctrl_ty()?)?,
         Opcode::UshrImm => binary(Value::ushr, arg(0)?, imm_as_ctrl_ty()?)?,
         Opcode::SshrImm => binary(Value::ishr, arg(0)?, imm_as_ctrl_ty()?)?,
-        Opcode::Bitrev => unimplemented!("Bitrev"),
-        Opcode::Clz => unimplemented!("Clz"),
-        Opcode::Cls => unimplemented!("Cls"),
-        Opcode::Ctz => unimplemented!("Ctz"),
-        Opcode::Popcnt => unimplemented!("Popcnt"),
+        Opcode::Bitrev => assign(Value::reverse_bits(arg(0)?)?),
+        Opcode::Clz => assign(arg(0)?.leading_zeros()?),
+        Opcode::Cls => {
+            let count = if Value::lt(&arg(0)?, &Value::int(0, ctrl_ty)?)? {
+                arg(0)?.leading_ones()?
+            } else {
+                arg(0)?.leading_zeros()?
+            };
+            assign(Value::sub(count, Value::int(1, ctrl_ty)?)?)
+        }
+        Opcode::Ctz => assign(arg(0)?.trailing_zeros()?),
+        Opcode::Popcnt => {
+            let count = if arg(0)?.ty().is_int() {
+                arg(0)?.count_ones()?
+            } else {
+                let lanes = extractlanes(&arg(0)?, ctrl_ty.lane_type())?
+                    .into_iter()
+                    .map(|lane| lane.count_ones().unwrap())
+                    .collect::<SimdVec<V>>();
```

Comment:
A really nice thing you can do with iterators is collecting from an iterator of `Result<T, _>` into `Result<Vec<T>, _>`. This fails on the first `Result` that isn't `Ok`, which is exactly what we want! I think the [rust book](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect) has a better explanation if you want to look more into it.

```suggestion
                    .map(|lane| lane.count_ones())
                    .collect::<ValueResult<SimdVec<V>>>()?;
```